pkg/daemon/daemon.go

```diff
@@ -131,57 +119,107 @@ func applyNodePTPProfile(pm *ProcessManager, nodeProfile *ptpv1.PtpProfile) erro
+	printWhenNotNil(nodeProfile.Phc2sysOpts, "Phc2sysOpts")
+	glog.Infof("------------------------------------")
+
 	if nodeProfile.Phc2sysOpts != nil {
-		pm.process = append(pm.process, &ptpProcess{
+		dn.processManager.process = append(dn.processManager.process, &ptpProcess{
 			name:   "phc2sys",
+			iface:  *nodeProfile.Interface,
 			exitCh: make(chan bool),
 			cmd:    phc2sysCreateCmd(nodeProfile)})
 	} else {
-		glog.Infof("applyNodePTPProfile: not starting phc2sys, phc2sysOpts is empty")
+		glog.Infof("applyNodePtpProfile: not starting phc2sys, phc2sysOpts is empty")
 	}
 
 	if nodeProfile.Ptp4lOpts != nil && nodeProfile.Interface != nil {
-		if nodeProfile.Ptp4lConf != nil && *nodeProfile.Ptp4lConf != "" {
-			if _, err := os.Stat(PTP4L_CONF_DIR); os.IsNotExist(err) {
-				err = os.Mkdir(PTP4L_CONF_DIR, 0644)
-				if err != nil {
-					return fmt.Errorf("failed to create the ptp4l custom configuration directory: %v", err)
-				}
-			}
-
-			err := ioutil.WriteFile(fmt.Sprintf("%s/%s.conf", PTP4L_CONF_DIR, *nodeProfile.Name), []byte(*nodeProfile.Ptp4lConf), 0644)
-			if err != nil {
-				return fmt.Errorf("failed to write the configuration file named %s: %v", fmt.Sprintf("%s/%s.conf", PTP4L_CONF_DIR, *nodeProfile.Name), err)
-			}
+		configPath := fmt.Sprintf("/var/run/ptp4l.%d.config", runID)
+		err := ioutil.WriteFile(configPath, []byte(*nodeProfile.Ptp4lConf), 0644)
+		if err != nil {
+			return fmt.Errorf("failed to write the configuration file named %s: %v", configPath, err)
 		}
 
-		pm.process = append(pm.process, &ptpProcess{
-			name:   "ptp4l",
-			exitCh: make(chan bool),
-			cmd:    ptp4lCreateCmd(nodeProfile)})
+		dn.processManager.process = append(dn.processManager.process, &ptpProcess{
+			name:            "ptp4l",
+			iface:           *nodeProfile.Interface,
+			ptp4lConfigPath: configPath,
+			ptp4lSocketPath: socketPath,
+			exitCh:          make(chan bool),
+			cmd:             ptp4lCreateCmd(nodeProfile, configPath)})
 	} else {
-		glog.Infof("applyNodePTPProfile: not starting ptp4l, ptp4lOpts or interface is empty")
+		glog.Infof("applyNodePtpProfile: not starting ptp4l, ptp4lOpts or interface is empty")
 	}
 
-	for _, p := range pm.process {
-		if p != nil {
-			time.Sleep(1 * time.Second)
-			go cmdRun(p)
-		}
-	}
 	return nil
 }
 
+func (dn *Daemon) addProfileConfig(runID int, nodeProfile *ptpv1.PtpProfile) string {
+	// TODO: later implement a merge capability
+	if nodeProfile.Ptp4lConf == nil || *nodeProfile.Ptp4lConf == "" {
+		// We need to copy this to another variable because is a pointer
+		config := string(dn.ptpUpdate.defaultPTP4lConfig)
+		nodeProfile.Ptp4lConf = &config
+	}
+
+	socketPath := fmt.Sprintf("/var/run/ptp4l.%d.socket", runID)
+
+	config := fmt.Sprintf("%s\nuds_address %s\nmessage_tag [%s]",
+		*nodeProfile.Ptp4lConf,
+		socketPath,
+		*nodeProfile.Interface)
+	nodeProfile.Ptp4lConf = &config
+
+	commandLine := fmt.Sprintf("%s -z %s -t [%s]",
```

Comment:
There is a warning in phc2sys man page:

```       Be  cautious when the same configuration file is used for both ptp4l and phc2sys.  Keep in mind, that values specified in the
       configuration file take precedence over their default values. If a certain option, which is common to ptp4l and  phc2sys,  is
       specified  to  a  non-default value in the configuration file (p.e., for ptp4l), then this non-default value applies also for
       phc2sys. This might be not what is expected.

       It is recommended to use seperate configuration files for ptp4l and phc2sys in order to avoid any unexpected behavior.
```

Since we are defining different config files for each ptp4l instance, shall we also consider using the same ptp4l config file for phc2sys instance that is created along with the ptp4l instance?

Reply:
If we use the same ptp4l.conf file for phc2sys and ptp4l instances created in one nodeProfile, then `-z` and `-t` options might not be needed. so this function can be used only for updating the ptp4l.conf.

Reply:
I took the implementation from the timemaster.

We are not pointing to the ptp4l config in the phc2sys program just to the socket.
Example:
```
/usr/sbin/ptp4l -m --summary_interval -4 -f nic1.conf -i eno1
/usr/sbin/phc2sys -a -r -m -u 0 -R 1.00 -z /var/run/ptp4l.0.socket  -t [0:eno1] -n 0

/usr/sbin/ptp4l -m --summary_interval -4 -f nic2.conf -i eno2
/usr/sbin/phc2sys -a -r -m -u 0 -R 1.00 -z /var/run/ptp4l.1.socket -t [0:eno2] -n 24
```

Reply:
Ack, so we are not using any ptp4l.conf file for phc2sys process.

Anther question: with multiple profiles be able to configured on one node, can we still always assume phc2sys and ptp4l come in pair? for example, will there be a profile that only contains phc2sys config `phc2sys -a -r -r`?

Reply:
I think we can assume that but the program support the use-case that only one process starts if the options for one of the process is empty the linuxptp-daemon will not start that process (this was already supported I didn't change that)